File: solutions/elixir/zipper/2/lib/zipper.ex
defmodule Zipper do
  @type t :: %Zipper{bin_tree: BinTree.t(), focus: BinTree.t(), path: t() | nil}

  defstruct [:bin_tree, :focus, :path]

  @doc """
  Get a zipper focused on the root node.
  """
  @spec from_tree(BinTree.t()) :: Zipper.t()
  def from_tree(%BinTree{} = bin_tree),
    do: %Zipper{
      bin_tree: bin_tree,
      focus: bin_tree,
      path: []
    }

  @doc """
  Get the complete tree from a zipper.
  """
  @spec to_tree(Zipper.t()) :: BinTree.t()
  def to_tree(%Zipper{bin_tree: bin_tree}),
    do: bin_tree

  @doc """
  Get the value of the focus node.
  """
  @spec value(Zipper.t()) :: any
  def value(%Zipper{
        focus: %BinTree{value: value}
      }),
      do: value

  @doc """
  Get the left child of the focus node, if any.
  """
  @spec left(Zipper.t()) :: Zipper.t() | nil
  def left(%Zipper{focus: %BinTree{left: nil}}),
    do: nil

  def left(%Zipper{
        bin_tree: bin_tree,
        focus: %BinTree{left: left},
        path: path
      }),
      do: %Zipper{
        bin_tree: bin_tree,
        focus: left,
        path: [:left | path]
      }

  @doc """
  Get the right child of the focus node, if any.
  """
  @spec right(Zipper.t()) :: Zipper.t() | nil
  def right(%Zipper{focus: %BinTree{right: nil}}),
    do: nil

  def right(%Zipper{
        bin_tree: bin_tree,
        focus: %BinTree{right: right},
        path: path
      }),
      do: %Zipper{
        bin_tree: bin_tree,
        focus: right,
        path: [:right | path]
      }

  @doc """
  Get the parent of the focus node, if any.
  """
  @spec up(Zipper.t()) :: Zipper.t() | nil
  def up(%Zipper{path: []}),
    do: nil

  def up(%Zipper{bin_tree: bin_tree, path: [_ | path_tail]}),
    do: %Zipper{
      bin_tree: bin_tree,
      focus:
        path_tail
        |> Enum.reverse()
        |> refocus(bin_tree),
      path: path_tail
    }

  defp refocus([], focus),
    do: focus

  defp refocus([:left | path_tail], focus),
    do: refocus(path_tail, focus, &left/1)

  defp refocus([:right | path_tail], focus),
    do: refocus(path_tail, focus, &right/1)

  defp refocus(path_tail, focus, focus_function) do
    %{focus: new_focus} =
      focus
      |> from_tree()
      |> focus_function.()

    refocus(path_tail, new_focus)
  end

  @doc """
  Set the value of the focus node.
  """
  @spec set_value(Zipper.t(), any) :: Zipper.t()
  def set_value(
        %Zipper{focus: old_focus, path: path} = zipper,
        new_value
      ),
      do: update_bin_tree(zipper, %BinTree{old_focus | value: new_value}, path)

  defp update_bin_tree(_zipper, %BinTree{} = new_focus, [] = path),
    do: %Zipper{bin_tree: new_focus, focus: new_focus, path: path}

  defp update_bin_tree(%Zipper{bin_tree: bin_tree} = zipper, new_focus, path),
    do: %Zipper{zipper | bin_tree: update_bin_tree(bin_tree, new_focus, path), focus: new_focus}

  defp update_bin_tree(%BinTree{} = bin_tree, new_focus, path),
    do: modify_bin_tree(%BinTree{} = bin_tree, new_focus, Enum.reverse(path))

  defp modify_bin_tree(%BinTree{} = _bin_tree, new_focus, [] = _path),
    do: new_focus

  defp modify_bin_tree(%BinTree{} = bin_tree, new_focus, [path_head | path_tail]),
    do:
      Map.put(
        bin_tree,
        path_head,
        modify_bin_tree(Map.fetch!(bin_tree, path_head), new_focus, path_tail)
      )

  @doc """
  Replace the left child tree of the focus node.
  """
  @spec set_left(Zipper.t(), BinTree.t() | nil) :: Zipper.t()
  def set_left(%Zipper{focus: focus, path: path} = zipper, nil),
    do: update_bin_tree(zipper, %{focus | left: nil}, path)

  def set_left(
        %Zipper{focus: focus, path: [_ | path_tail]} = zipper,
        %BinTree{value: _value, left: nil, right: nil} = leaf
      ),
      do: update_bin_tree(zipper, %{focus | left: leaf}, [:left | path_tail])

  def set_left(
        %Zipper{bin_tree: bin_tree} = zipper,
        %BinTree{value: _value, left: _left, right: _right} = subtree
      ),
      do: %Zipper{
        zipper
        | bin_tree: %BinTree{bin_tree | left: subtree},
          focus: subtree
      }

  @doc """
  Replace the right child tree of the focus node.
  """
  @spec set_right(Zipper.t(), BinTree.t() | nil) :: Zipper.t()
  def set_right(%Zipper{focus: focus, path: path} = zipper, nil),
    do: update_bin_tree(zipper, %{focus | right: nil}, path)

  def set_right(
        %Zipper{focus: focus, path: [_ | path_tail]} = zipper,
        %BinTree{value: _value, left: nil, right: nil} = leaf
      ),
      do: update_bin_tree(zipper, %{focus | right: leaf}, [:right | path_tail])

  def set_right(%Zipper{bin_tree: bin_tree} = zipper, right),
    do: %Zipper{zipper | bin_tree: %BinTree{bin_tree | right: right}, focus: right}
end

File: solutions/ocaml/zipper/1/zipper.ml
type 'a t = {
  focus: 'a Tree.t;        (* the focused tree node *)
  parent: 'a t option;     (* parent zipper *)
  is_left: bool;           (* whether focus is left child of parent *)
}

let rec equal (value_eq : 'a -> 'a -> bool) (z1 : 'a t) (z2 : 'a t) : bool =
  Tree.equal value_eq z1.focus z2.focus &&
  Option.equal (equal value_eq) z1.parent z2.parent &&
  z1.is_left = z2.is_left

let t_of_sexp (_ : Sexplib0.Sexp.t -> 'a) (_ : Sexplib0.Sexp.t) : 'a t =
  failwith "t_of_sexp not needed"

let sexp_of_t (_ : 'a -> Sexplib0.Sexp.t) (_ : 'a t) : Sexplib0.Sexp.t =
  Sexplib0.Sexp.Atom "<zipper>"

(* Create zipper focused on root *)
let of_tree (tree: 'a Tree.t) : 'a t =
  { focus = tree; parent = None; is_left = false }

(* Get complete tree from zipper *)
let rec to_tree (zipper: 'a t) : 'a Tree.t =
  match zipper.parent with
  | None -> zipper.focus
  | Some p ->
      let updated_parent =
        if zipper.is_left then
          { p.focus with left = Some zipper.focus }
        else
          { p.focus with right = Some zipper.focus }
      in
      to_tree { p with focus = updated_parent }

(* Get value of focus *)
let value (zipper: 'a t) : 'a =
  zipper.focus.value

(* Get left child zipper *)
let left (zipper: 'a t) : 'a t option =
  match zipper.focus.left with
  | None -> None
  | Some l -> Some { focus = l; parent = Some zipper; is_left = true }

(* Get right child zipper *)
let right (zipper: 'a t) : 'a t option =
  match zipper.focus.right with
  | None -> None
  | Some r -> Some { focus = r; parent = Some zipper; is_left = false }

(* Get parent zipper *)
let up (zipper: 'a t) : 'a t option =
  zipper.parent

(* Set value of focus *)
let set_value (new_value: 'a) (zipper: 'a t) : 'a t =
  { zipper with focus = { zipper.focus with value = new_value } }

(* Set left child of focus *)
let set_left (tree_opt: 'a Tree.t option) (zipper: 'a t) : 'a t =
  { zipper with focus = { zipper.focus with left = tree_opt } }

(* Set right child of focus *)
let set_right (tree_opt: 'a Tree.t option) (zipper: 'a t) : 'a t =
  { zipper with focus = { zipper.focus with right = tree_opt } }
